boa_engine/src/builtins/temporal/date_equations.rs

```diff
@@ -0,0 +1,136 @@
+//! This file represents all equations listed under section 13.4 of the [Temporal Specification][spec]
+//!
+//! [spec]: https://tc39.es/proposal-temporal/#sec-date-equations
+
+use std::ops::Mul;
+
+pub(crate) fn epoch_time_to_day_number(t: f64) -> i32 {
+    (t / super::NS_PER_DAY as f64).floor() as i32
+}
+
+pub(crate) fn mathematical_days_in_year(y: i32) -> i32 {
+    if y % 4 != 0 {
+        365
+    } else if y % 4 == 0 && y % 100 != 0 {
+        366
+    } else if y % 100 == 0 && y % 400 != 0 {
+        365
+    } else {
+        // Assert that y is divisble by 400 to ensure we are returning the correct result.
+        assert_eq!(y % 400, 0);
+        366
+    }
+}
+
+pub(crate) const fn epoch_day_number_for_year(y: i32) -> i32 {
+    365 * (y - 1970) + ((y - 1970) / 4) - ((y - 1901) / 100) + ((y - 1601) / 400)
+}
+
+// TODO: potentially inaccurate -> Need to further test this and epoch_day_number_for_year
+pub(crate) fn epoch_time_for_year(y: i32) -> f64 {
+    super::NS_PER_DAY as f64 * f64::from(epoch_day_number_for_year(y))
+}
+
+pub(crate) fn epoch_time_to_epoch_year(t: f64) -> i32 {
+    // roughly calculate the largest possible year given the time t,
+    // then check and refine the year.
+    let day_count = epoch_time_to_day_number(t);
+    let mut year = day_count / 365;
+    loop {
+        if epoch_time_for_year(year) <= t {
+            break;
+        }
+        year -= 1;
+    }
+
+    year
+}
+
+/// Returns either 1 (true) or 0 (false)
+pub(crate) fn mathematical_in_leap_year(t: f64) -> i32 {
+    mathematical_days_in_year(epoch_time_to_epoch_year(t)) - 365
+}
+
+pub(crate) fn epoch_time_to_month_in_year(t: f64) -> i32 {
+    let days = epoch_time_to_day_in_year(t);
+    let in_leap_year = mathematical_in_leap_year(t) == 1;
+
+    match days {
+        0..=30 => 0,
+        31..=59 if in_leap_year => 1,
+        31..=58 => 1,
+        60..=90 if in_leap_year => 2,
+        59..=89 => 2,
+        91..=121 if in_leap_year => 3,
+        90..=120 => 3,
+        122..=151 if in_leap_year => 4,
+        121..=150 => 4,
+        152..=182 if in_leap_year => 5,
+        151..=181 => 5,
+        183..=213 if in_leap_year => 6,
+        182..=212 => 6,
+        214..=243 if in_leap_year => 7,
+        213..=242 => 7,
+        244..=273 if in_leap_year => 8,
+        243..=272 => 8,
+        274..=304 if in_leap_year => 9,
+        273..=303 => 9,
+        305..=334 if in_leap_year => 10,
+        304..=333 => 10,
+        335..=366 if in_leap_year => 11,
+        334..=365 => 11,
+        _ => unreachable!(),
+    }
```

Comment:
This could benefit from representing it as two arrays instead of a set of matches. That way, you can just use `binary_search` to find the position where `days` is less or equal than the value in the array:
```Rust
const DAYS: [u8; 12] = [30, 58, 89, ...];

let month = DAYS.binary_search(&days); // then unify the `Err` case with the `Ok` case. Both should always return the same index anyways.
```